flink-tests/src/test/java/org/apache/flink/test/scheduling/ReactiveModeITCase.java

```diff
@@ -0,0 +1,232 @@
+        input.addSink(new ParallelismTrackingSink<>()).getTransformation();
+
+        ParallelismTrackingSource.expectInstances(
+                NUMBER_SLOTS_PER_TASK_MANAGER * INITIAL_NUMBER_TASK_MANAGERS);
+        ParallelismTrackingSink.expectInstances(
+                NUMBER_SLOTS_PER_TASK_MANAGER * INITIAL_NUMBER_TASK_MANAGERS);
+
+        env.executeAsync();
+
+        ParallelismTrackingSource.waitForInstances();
+        ParallelismTrackingSink.waitForInstances();
+
+        // expect scale up to 2 TaskManagers:
+        ParallelismTrackingSource.expectInstances(NUMBER_SLOTS_PER_TASK_MANAGER * 2);
+        ParallelismTrackingSink.expectInstances(NUMBER_SLOTS_PER_TASK_MANAGER * 2);
+
+        miniClusterResource.getMiniCluster().startTaskManager();
+
+        ParallelismTrackingSource.waitForInstances();
+        ParallelismTrackingSink.waitForInstances();
+    }
+
+    @Test
+    public void testScaleDownOnTaskManagerLoss() throws Exception {
+        // test preparation: ensure we have 2 TaskManagers running
+        startAdditionalTaskManager();
+
+        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+        env.setRestartStrategy(RestartStrategies.fixedDelayRestart(Integer.MAX_VALUE, 0L));
+        final DataStream<String> input = env.addSource(new ParallelismTrackingSource());
+        input.addSink(new ParallelismTrackingSink<>()).getTransformation();
+
+        ParallelismTrackingSource.expectInstances(NUMBER_SLOTS_PER_TASK_MANAGER * 2);
+        ParallelismTrackingSink.expectInstances(NUMBER_SLOTS_PER_TASK_MANAGER * 2);
+
+        env.executeAsync();
+
+        ParallelismTrackingSource.waitForInstances();
+        ParallelismTrackingSink.waitForInstances();
+
+        // scale down to 1 TaskManagers:
+        ParallelismTrackingSource.expectInstances(NUMBER_SLOTS_PER_TASK_MANAGER);
+        ParallelismTrackingSink.expectInstances(NUMBER_SLOTS_PER_TASK_MANAGER);
+
+        miniClusterResource.getMiniCluster().terminateTaskManager(0).get();
+
+        ParallelismTrackingSource.waitForInstances();
+        ParallelismTrackingSink.waitForInstances();
+    }
+
+    private int getNumberOfConnectedTaskManagers() throws ExecutionException, InterruptedException {
+        return miniClusterResource
+                .getMiniCluster()
+                .requestClusterOverview()
+                .get()
+                .getNumTaskManagersConnected();
+    }
+
+    private void startAdditionalTaskManager() throws Exception {
+        miniClusterResource.getMiniCluster().startTaskManager();
+        CommonTestUtils.waitUntilCondition(
+                () -> getNumberOfConnectedTaskManagers() == 2,
+                Deadline.fromNow(Duration.ofMillis(10_000L)));
+    }
+
+    private static class ParallelismTrackingSource implements ParallelSourceFunction<String> {
+        private volatile boolean running = true;
+
+        @GuardedBy("instances")
+        private static CountDownLatch instances;
```

Comment:
I don't think that you should use a non-final field as a lock. I would suggest to introduce a final lock. Moreover we want to synchronize writes/reads to `instances` which does not work if you use `instances` as the lock.